### PR TITLE
If resolvedFileName differs with realPath only in casing use the resolvedFileName before realpath so that errors can be reported with forceConsistentCasingInFileNames

### LINKED ISSUE #49470: `forceConsistentCasingInFileNames` doesn't work for modules in `node_modules`
# Bug Report

### 🔎 Search Terms

- `forceConsistentCasingInFileNames`
- `node_modules`

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about `forceConsistentCasingInFileNames`

### ⏯ Playground Link

Not available—requires `node_modules` folder

### 💻 Code

With `forceConsistentCasingInFileNames` enabled and `fp-ts` installed in `node_modules`:

`src/struct.d.ts`:
```ts
export const foo = 1;

```

```ts
// Expected error, but got none ❌
import * as xs1 from "fp-ts/lib/Struct";
// Expected error, but got none ❌
import * as xs2 from "fp-ts/lib/struct";
// Error as expected ✅
import * as xs3 from "./Struct";
// Error as expected ✅
import * as xs4 from "./struct";
```

### 🙁 Actual behavior

See code comments above.

### 🙂 Expected behavior

See code comments above.

### PR DESCRIPTION
**Pull request by sheetalkamat**
Fixes #49470

### CODE REVIEW

**Review by andrewbranch — APPROVED** [review]

File: src/compiler/moduleNameResolver.ts [hunk 1]
```diff
@@ -392,10 +392,11 @@ namespace ts {
         if (resolved) {
             const { fileName, packageId } = resolved;
             const resolvedFileName = options.preserveSymlinks ? fileName : realPath(fileName, host, traceEnabled);
+            const pathsAreEqual = arePathsEqual(fileName, resolvedFileName, host);
             resolvedTypeReferenceDirective = {
                 primary,
-                resolvedFileName,
-                originalPath: arePathsEqual(fileName, resolvedFileName, host) ? undefined : fileName,
+                resolvedFileName: pathsAreEqual ? fileName : resolvedFileName,
```
Comment (amcasey) [inline src/compiler/moduleNameResolver.ts:398]: Personally, I think this needs a comment so someone doesn't simplify it back to `resolvedFileName`.

**Review by amcasey — APPROVED** [review]

File: src/compiler/moduleNameResolver.ts (hunk 1 above)
Comment (amcasey) [reply src/compiler/moduleNameResolver.ts:399]: Perfect - thanks!

**Comment by sheetalkamat** [comment]
@DanielRosenwasser do you think we can take this as well given we are taking the other fix for forceConsistentCasingInFileNames. 

@typescript-bot cherry-pick this to release-4.8

**Comment by amcasey** [comment]
My personal opinion is that this is less important than the long-path fix, but it seems low-risk and I have no particular objection.

**Comment by imaginetheheadline** [comment]
@sheetalkamat @andrewbranch @amcasey This PR causes a performance regression with eslint in our code base. I have narrowed this down by testing our code with specific TS commits, and this is the culprit.

We have: 
- 288,000 lines of TS
- Webstorm
  - Running `eslint --fix` on save
- ESLint config has prettier plugin which reformats code when `eslint --fix` runs

All of these together means we went from <0.5s to run `eslint --fix` on save to now taking ~10 seconds for each save.

Can you suggest a mechanism by which this change could have contributed to this slowdown? I haven't so far dived deeply into this code.

**Comment by amcasey** [comment]
@imaginetheheadline Thanks for the heads up!  No, it's definitely not obvious to me how this could cause a major slowdown.  We've had problems in the past with accidentally making too many `realpath` calls, but it's not clear to me how this could cause that.

I think that if you grab a trace with [pprof-it](https://github.com/jakebailey/pprof-it), the cause will jump right out.  The pprof file will contain only information about the code that's executing, not the code that's being analyzed, so it should be safe to share (though it _will_ contain the paths to your copies of the code being run (e.g. c:\users\me\typescript\lib\tsc.js)).

**Comment by imaginetheheadline** [comment]
@amcasey Have been running `pprof-it` a little and so far I'm noticing the following.

Context: I've been running a locally-installed TypeScript starting from this commit, and then comparing with removing this commit, and see the following results:

**Pre-commit**
- `eslint --fix` runs in about 48 seconds, with garbage collection taking an additional 9-12 seconds.

**Post-commit**
- `eslint --fix` runs in about 77 seconds, with garbage collection taking an additional 65-75 seconds
- The main increase I can see so far, is time to run `createDiagnosticExplainingFile`
  - I have logged the `fileProcessingReason` here and can see that `FileIncludeKind.Import` is now at ~22,000 compared to ~6,500. Other `FileIncludeKind` values are about the same.
- Memory usage in `createDiagnosticExplainingFile` is also high now.

Screen Shot 2022-09-19 at 12 42 17 pm

  
If anything stands out from that that you can recommend I check next, let me know. Otherwise, I may be able to share the `pprof` files as well.

**Comment by imaginetheheadline** [comment]
@amcasey Yeah, sure, I'll raise a proper issue tomorrow.

The diagnostics is a good question. Can you clarify what diagnostics means in this case? Would you expect that to be always running across the code, or would we only want that to be turned on in particular instances? I had checked our `tsconfig` to see if we were turning something on but I didn't see it. I'm also wondering if this might be the result of the config of some of the chain of tools here: typescript -> typescript-eslint -> eslint -> prettier.

I'll put all of this on the reported issue tomorrow.

**Comment by amcasey** [comment]
@imaginetheheadline Oh, sorry, diagnostics is jargon for the error-like things (errors, warnings, and, in some languages, suggestions).  Broadly, things that might display as a squiggle in the editor.  In a general sense, it can refer either the error check or a specific instance of an error, but I was referring to instances of errors.  That is, if you looked in your Errors/Problems window, your output suggests you would see 6500 entries (now 22000)?

**Comment by imaginetheheadline** [comment]
@amcasey Yeah I see. It says 55 errors, 2195 warnings, and 13,170 typos lol. Those don't seem to be all from TypeScript though. I'll restart the conversation with more info when I file the issue. Thanks so much for your replies!

**Comment by DanielRosenwasser** [comment]
If I'm not mistaken, those diagnostics are not error diagnostics - they're generated when running with things like `explainFiles`, `traceResolution`, and more.

I could be wrong though.

**Pull request merged**